src/core/pdf/popplerapi/PopplerGlibPage.cpp

```diff
@@ -78,3 +88,194 @@ auto PopplerGlibPage::findText(std::string& text) -> std::vector<XojPdfRectangle
+        }
+        char* textBytes = poppler_page_get_text_for_area(page, &pRect);
+        g_assert_nonnull(textBytes);
+
+        double y = rectArray[0].y2;
+        std::ostringstream ss;
+        for (guint i = 0; i < numRects; i++) {
+            // do not copy characters whose bounding box has a non-empty intersection with rect
+            const auto& r = rectArray[i];
+            {
+                auto x1 = std::max(rect.x1, r.x1);
+                auto y1 = std::max(rect.y1, r.y1);
+                auto x2 = std::min(rect.x2, r.x2);
+                auto y2 = std::min(rect.y2, r.y2);
+
+                bool inBounds = x2 > x1 && y2 > y1;
+                if (!inBounds)
+                    continue;
+            }
+
+            const auto eps = 1e-5;
+            if (std::abs(y - r.y2) > eps) {
+                // new line
+                ss << '\n';
+                y = rectArray[i].y2;
+            }
+
+            char* const startPos = g_utf8_offset_to_pointer(textBytes, i);
+            char* const endPos = g_utf8_offset_to_pointer(textBytes, i + 1);
+            for (long j = 0; j < static_cast<ptrdiff_t>(endPos - startPos); ++j) { ss << startPos[j]; }
+        }
+        g_free(textBytes);
+        return ss.str();
+    } else {
+        char* text = poppler_page_get_selected_text(page, pStyle, &pRect);
+        if (text) {
+            std::string ret(text);
+            g_free(text);
+            return ret;
+        } else {
+            return "";
+        }
+    }
+}
+
+auto PopplerGlibPage::selectTextRegion(const XojPdfRectangle& rect, XojPdfPageSelectionStyle style) -> cairo_region_t* {
+    PopplerRectangle pRect = {rect.x1, rect.y1, rect.x2, rect.y2};
+    const auto pStyle = getPopplerSelectionStyle(style);
+    // The computed region is technically wrong for
+    // XojPdfPageSelectionStyle::Area, but there is no selection preview with
+    // area select.
+    cairo_region_t* region = poppler_page_get_selected_region(page, 1.0, pStyle, &pRect);
+    return region;
+}
+
+namespace {
+cairo_rectangle_int_t cairoRectFromDouble(double x1, double y1, double width, double height) {
+    return {static_cast<int>(x1), static_cast<int>(y1), static_cast<int>(width), static_cast<int>(height)};
+}
+}  // namespace
+
+auto PopplerGlibPage::selectTextLines(const XojPdfRectangle& rect, XojPdfPageSelectionStyle style) -> TextSelection {
+    std::vector<XojPdfRectangle> textRects;
+
+    PopplerRectangle* rectArray = nullptr;
+    guint numRects = 0;
+    if (style == XojPdfPageSelectionStyle::Area) {
+        PopplerRectangle area{rect.x1, rect.y1, rect.x2, rect.y2};
+        if (!poppler_page_get_text_layout_for_area(this->page, &area, &rectArray, &numRects)) {
+            return {.region = cairo_region_create(), .rects = textRects};
+        }
+    } else {
+        if (!poppler_page_get_text_layout(this->page, &rectArray, &numRects)) {
+            return {.region = cairo_region_create(), .rects = textRects};
```

Comment:
This is c++20 code and not allowed currently

Reply:
Both gcc and clang support the C99 style designated initializers, and this code has already been merged.